java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,761 @@
+    currentBlock = createBranch(s, loopback, falseBranch);
+    buildCondition(s.condition(), loopback, falseBranch);
+    // process body
+    currentBlock = createBlock(currentBlock);
+    continueTargets.addLast(loopback);
+    breakTargets.addLast(falseBranch);
+    build(s.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    loopback.successors.add(currentBlock);
+    currentBlock = createBlock(currentBlock);
+  }
+
+  private void buildForEachStatement(ForEachStatement tree) {
+    // TODO(npe) One solution is to create a forstatement node depending on type of expression (iterable or array) and build CFG from
+    // it.
+    Block afterLoop = currentBlock;
+    currentBlock = createBlock();
+    Block loopback = currentBlock;
+    continueTargets.addLast(loopback);
+    breakTargets.addLast(afterLoop);
+    build(tree.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    currentBlock = createBranch(tree, currentBlock, afterLoop);
+    loopback.successors.add(currentBlock);
+    build(tree.variable());
+    build(tree.expression());
+    currentBlock = createBlock(currentBlock);
+  }
+
+  private void buildForStatement(ForStatementTree tree) {
+    Block falseBranch = currentBlock;
+    // process step
+    currentBlock = createBlock();
+    Block updateBlock = currentBlock;
+    for (StatementTree updateTree : Lists.reverse(tree.update())) {
+      build(updateTree);
+    }
+    continueTargets.addLast(currentBlock);
+    // process body
+    currentBlock = createBlock(currentBlock);
+    breakTargets.addLast(falseBranch);
+    build(tree.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    Block body = currentBlock;
+    // process condition
+    ExpressionTree condition = tree.condition();
+    if (condition != null) {
+      currentBlock = createBranch(tree, body, falseBranch);
+      buildCondition(condition, body, falseBranch);
+    } else {
+      currentBlock = createUnconditionalJump(tree, body);
+    }
+    updateBlock.successors.add(currentBlock);
+    // process init
+    currentBlock = createBlock(currentBlock);
+    for (StatementTree init : Lists.reverse(tree.initializer())) {
+      build(init);
+    }
+  }
+
+  private void buildTryStatement(TryStatementTree tree) {
+    // FIXME only path with no failure constructed for now, (not taking try with resources into consideration).
+    TryStatementTree tryStatementTree = tree;
+    currentBlock = createBlock(currentBlock);
+    BlockTree finallyBlock = tryStatementTree.finallyBlock();
+    if (finallyBlock != null) {
+      build(finallyBlock);
+    }
+    currentBlock = createBlock(currentBlock);
+    build(tryStatementTree.block());
+    currentBlock = createBlock(currentBlock);
+    currentBlock.elements.add(tree);
+  }
+
+  private void buildThrowStatement(ThrowStatementTree tree) {
+    // FIXME this won't work if it is intended to be caught by a try statement.
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)